### ISSUE TITLE
Unsafe folder error when committing

### ISSUE THREAD

**Issue by cuppa-t-lager** [magit/magit#3148]
Hi...

Started to get issues with Magit this morning. I'm running on OSX, using GNU Emacs 25.2.1 (x86_64-apple-darwin15.0.0, Carbon Version 157 AppKit 1404.13)
 of 2017-08-05

When trying to commit I receive the following error. Generally after a restart  I am able to commit. 

Any Ideas?

Debugger entered--Lisp error: (error "The directory ‘/var/folders/k0/21tqv_bn1sq0fjjmwzt11n0c0000gn/T//emacs501’ is unsafe")
  signal(error ("The directory ‘/var/folders/k0/21tqv_bn1sq0fjjmwzt11n0c0000gn/T//emacs501’ is unsafe"))
  error("The directory `%s' is unsafe" "/var/folders/k0/21tqv_bn1sq0fjjmwzt11n0c0000gn/T//emacs501")
  server-ensure-safe-dir("/var/folders/k0/21tqv_bn1sq0fjjmwzt11n0c0000gn/T//emacs501")
  server-start()
  with-editor--setup()
  magit-run-git-with-editor("commit" ("--"))
  magit-commit(nil)
  funcall-interactively(magit-commit nil)
  call-interactively(magit-commit)
  magit-invoke-popup-action(99)
  funcall-interactively(magit-invoke-popup-action 99)
  call-interactively(magit-invoke-popup-action nil nil)
  command-execute(magit-invoke-popup-action)

**Comment by npostavs** [comment]
What's the result of evaluating:
`(file-attributes "/var/folders/k0/21tqv_bn1sq0fjjmwzt11n0c0000gn/T//emacs501" 'integer)`
`(file-modes "/var/folders/k0/21tqv_bn1sq0fjjmwzt11n0c0000gn/T//emacs501")`
`(user-uid)`

**Comment by cuppa-t-lager** [comment]
(t 6 501 20 (22930 64534 0 0) (22930 64534 0 0) (22930 64534 0 0) 204 "drwxr-xr-x" t 86284678 16777220)
493 (#o755, #x1ed, ?ǭ)
and 
501 (#o765, #x1f5, ?ǵ)

**Comment by npostavs** [comment]
Okay, so you have `(file-modes "/var/folders/k0/21tqv_bn1sq0fjjmwzt11n0c0000gn/T//emacs501")` => `#o755` which means that it's world readable; this is considered unsafe by `server-ensure-safe-dir`, i.e., this check at the bottom fails: `(zerop (logand ?\077 (file-modes dir)))`.

It seems that the umask isn't being set right, do you get `#o755` from this too?
```elisp
(cl-letf (((default-file-modes) ?\700)
          (dir (expand-file-name "test-dir" temporary-file-directory)))
  (make-directory dir t)
  (prog1 (file-modes dir)
    (delete-directory dir)))
```

**Comment by cuppa-t-lager** [comment]
Ive restarted so to continue working. The results from your function are now: 448 (#o700, #x1c0, ?ǀ). Which I presume is in keeping. Its probably understanding whats switching it to #o755 whilst I'm working.

**Comment by npostavs** [comment]
Right, the question is, is it a one-off that the directory gets created with #o755, or is there some bad state you somehow land in where every directory gets those permissions from then on.

**Comment by npostavs** [comment]
Perhaps you can have strace (or whatever the macOS equivalent is, dtrace? dtruss?) running on Emacs to check `umask` calls, and see if there are some unbalanced ones when it breaks again.

**Comment by cuppa-t-lager** [comment]
Hi... has happened again after behaving well all day today. I'm not too technical so can you "fly me in" what I need to do to get you the information you require. Will probably have to reboot for now... but will be ready for the next time this breaks again... thanks

**Comment by npostavs** [comment]
If it happens again, try evaluating the expression at https://github.com/magit/magit/issues/3148#issuecomment-322498007 (i.e., copy it to `*scratch*` and then hit `C-M-x` with point sitting on it) before you restart.

I don't have a macOS box to test this, but try running `dtrace -n emacs -t umask` in a terminal, and leave it running as you use Emacs normally. It may give us some clues for when you hit the problem again.

**Comment by cuppa-t-lager** [comment]
OK were back:

Debugger entered--Lisp error: (error "The directory ‘/var/folders/k0/21tqv_bn1sq0fjjmwzt11n0c0000gn/T//emacs501’ is unsafe")
  signal(error ("The directory ‘/var/folders/k0/21tqv_bn1sq0fjjmwzt11n0c0000gn/T//emacs501’ is unsafe"))
  error("The directory `%s' is unsafe" "/var/folders/k0/21tqv_bn1sq0fjjmwzt11n0c0000gn/T//emacs501")
  server-ensure-safe-dir("/var/folders/k0/21tqv_bn1sq0fjjmwzt11n0c0000gn/T//emacs501")
  server-start()
  with-editor--setup()
  magit-run-git-with-editor("commit" ("--"))
  magit-commit(nil)
  funcall-interactively(magit-commit nil)
  call-interactively(magit-commit)
  magit-invoke-popup-action(99)
  funcall-interactively(magit-invoke-popup-action 99)
  call-interactively(magit-invoke-popup-action nil nil)
  command-execute(magit-invoke-popup-action)

Runnunning #3148 (comment) returns: 448 (#o700, #x1c0, ?ǀ)

**Comment by npostavs** [comment]
> Runnunning #3148 (comment) returns: 448 (#o700, #x1c0, ?ǀ)

Hmm, so the umask mechanism seems to be still working. If you continue from the error, and try committing again, does it still fail?

> On my mac: dtrace does not have a -t flag... "dtrace -n emacs -t umask" is there another trace I could use?

Um, there seems to be several things called `dtrace` and I may have got confused. I think that should have been `dtruss -n emacs -t umask`.

https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man1/dtruss.1m.html

**Comment by tarsius** [comment]
Closing since this most likely is not a Magit issue.
